### ISSUE TITLE
How to bind a nested class property to the attribute of html element with GetEncapsulatedData Method?

### ISSUE THREAD

**Issue by rwecho** [zzzprojects/html-agility-pack#505]
https://github.com/zzzprojects/html-agility-pack/blob/c12580b2025144d618d3397dec1fd81132e3f822/src/HtmlAgilityPack.Shared/HtmlNode.Encapsulator.cs#L298

In most situations， I like to use XPath attribute binding to a Property. But in the nested class, it can not access the parent node, the parent node's attribute often records some important information, like id.

``` csharp

   [HasXPath]
    class TestA
    {
        [XPath("a")]
        public List<Item> Items { get; set; }
        [HasXPath]
        public class Item
        {
            [XPath(".", "href")] //not working
            [SkipNodeNotFound]
            public string Href { get; set; }

            [XPath(".")]
            [SkipNodeNotFound]
            public string Name { get; set; }
        }

    }

    var html = @"
<div>
<div>Hello
<a href='1.html'>1.html</a>
<a href='2.html'>2.html</a>
</div>
<div>World</div>
</div>
";

     var document = new HtmlDocument();
     Document. LoadHtml(html);
     var testA = Document. DocumentNode.GetEncapsulatedData<TestA>();

``` 

In the HtmlNode.Encapsulator.cs#L298,  Is using OuterHtml better than InnerHtml?

**Comment by elgonzo** [comment]
> In the HtmlNode.Encapsulator.cs#L298, Is using OuterHtml better than InnerHtml?

The bug seems to be not adhering to the `[XPath]` attribute's NodeReturnType value here (the NodeReturnType value is supposed to govern whether InnerText, InnerHtml or OuterHtml is used for "encapsulation".

HtmlNode.Encapsulator.cs#L298 isn't the only place exhibiting this issue,

https://github.com/zzzprojects/html-agility-pack/blob/c12580b2025144d618d3397dec1fd81132e3f822/src/HtmlAgilityPack.Shared/HtmlNode.Encapsulator.cs#L298

HtmlNode.Encapsulator.cs#L180 also forgets to take HasXPathAttribute.NodeReturnType  into account:
https://github.com/zzzprojects/html-agility-pack/blob/c12580b2025144d618d3397dec1fd81132e3f822/src/HtmlAgilityPack.Shared/HtmlNode.Encapsulator.cs#L180

**Comment by elgonzo** [comment]
Side note: There is a related issue that also warrants fixing when the proper NodeReturnType support is being implemented. Currently, when a property of a simple type like string has a `[XPath]` attribute with an invalid NodeReturnType value, a bare-bones and non-descript System.Exception is being thrown:

https://github.com/zzzprojects/html-agility-pack/blob/c12580b2025144d618d3397dec1fd81132e3f822/src/HtmlAgilityPack.Shared/HtmlNode.Encapsulator.cs#L629

This should to be changed into a meaningful exception with a message that informs about the _concrete_ cause of the exception.

**Comment by rwecho** [comment]
@elgonzo I created a PT on your advice, please help me review it.

**Comment by JonathanMagnan** [comment]
Hello @rwecho ,

We will close this issue has your PR has been merged and now available in the v1.11.50

Best Regards,

Jon
